security/pkg/server/ca/node_auth.go

```diff
@@ -15,28 +15,114 @@
 // valid identities which run on that node (rather than arbitrary ones).
-type NodeAuthorizer struct {
-	trustedNodeAccounts map[types.NamespacedName]struct{}
+type NodeAuthorizer interface {
+	authenticateImpersonation(ctx context.Context, caller security.KubernetesInfo, requestedIdentityString string) error
+}
+
+// MulticlusterNodeAuthorizor is an implementation of NodeAuthorizer for multi-cluster environmnets.
+// please refer to https://docs.google.com/document/d/10uf4EvUVif4xGeCYQydaKh9Yaz9wpysao7gyLewJY2Q.
+// It is responsible for maintaining an index of node authenticators, one per cluster.
+// node authorizations from one cluster will be forwarded to the node authorizer for the same cluster.
+type MulticlusterNodeAuthorizor struct {
+	remoteNodeAuthenticators map[cluster.ID]*ClusterNodeAuthorizer
+	m                        sync.Mutex
+	filter                   namespace.DiscoveryFilter
+	trustedNodeAccounts      sets.Set[types.NamespacedName]
+}
+
+func NewMulticlusterNodeAuthenticator(filter namespace.DiscoveryFilter, trustedNodeAccounts sets.Set[types.NamespacedName],
+	addClusterHandler func(multicluster.ClusterHandler),
+) *MulticlusterNodeAuthorizor {
+	m := &MulticlusterNodeAuthorizor{
+		remoteNodeAuthenticators: map[cluster.ID]*ClusterNodeAuthorizer{},
+		filter:                   filter,
+		trustedNodeAccounts:      trustedNodeAccounts,
+	}
+	if addClusterHandler != nil {
+		addClusterHandler(m)
+	}
+	return m
+}
+
+func (mNa *MulticlusterNodeAuthorizor) ClusterAdded(cluster *multicluster.Cluster, _ <-chan struct{}) {
+	na, err := NewClusterNodeAuthorizer(cluster.Client, mNa.filter, mNa.trustedNodeAccounts)
+	if err != nil {
+		serverCaLog.Errorf("failed to initialize node authorizer for cluster %v: %v", cluster.ID, err)
+		return
+	}
+	mNa.m.Lock()
+	defer mNa.m.Unlock()
+	mNa.addCluster(cluster.ID, na)
+}
+
+func (mNa *MulticlusterNodeAuthorizor) ClusterUpdated(cluster *multicluster.Cluster, stop <-chan struct{}) {
+	na, err := NewClusterNodeAuthorizer(cluster.Client, mNa.filter, mNa.trustedNodeAccounts)
+	if err != nil {
+		serverCaLog.Errorf("failed to initialize node authorizer for cluster %v: %v", cluster.ID, err)
+	}
+	mNa.m.Lock()
+	defer mNa.m.Unlock()
+	mNa.deleteCluster(cluster.ID)
+	mNa.addCluster(cluster.ID, na)
+}
+
+func (mNa *MulticlusterNodeAuthorizor) ClusterDeleted(key cluster.ID) {
+	mNa.m.Lock()
+	defer mNa.m.Unlock()
+	mNa.deleteCluster(key)
+}
+
+func (mNa *MulticlusterNodeAuthorizor) addCluster(clusterID cluster.ID, na *ClusterNodeAuthorizer) {
+	mNa.remoteNodeAuthenticators[clusterID] = na
+}
+
+func (mNa *MulticlusterNodeAuthorizor) deleteCluster(clusterID cluster.ID) {
+	delete(mNa.remoteNodeAuthenticators, clusterID)
+}
+
+func (mNa *MulticlusterNodeAuthorizor) authenticateImpersonation(ctx context.Context, caller security.KubernetesInfo, requestedIdentityString string) error {
+	clusterID := kubeauth.ExtractClusterID(ctx)
+	na, found := mNa.remoteNodeAuthenticators[clusterID]
```

Comment:
how do you protect it with lock?

Reply:
Thank you for pointing that out. I have now used m.m (a `sync.Mutex`)to protect accessing to the remoteNodeAuthorizer here.